crates/typst-library/src/visualize/gradient.rs

```diff
@@ -1284,25 +1284,29 @@ fn process_stops(stops: &[Spanned<GradientStop>]) -> SourceResult<Vec<(Color, Ra
 
 /// Sample the stops at a given position.
 fn sample_stops(stops: &[(Color, Ratio)], mixing_space: ColorSpace, t: f64) -> Color {
+    assert!(!stops.is_empty(), "Cannot sample from gradient with no stops");
+
     let t = t.clamp(0.0, 1.0);
-    let mut low = 0;
-    let mut high = stops.len();
+    let mut j = stops.partition_point(|(_, ratio)| ratio.get() < t);
 
-    while low < high {
-        let mid = (low + high) / 2;
-        if stops[mid].1.get() < t {
-            low = mid + 1;
-        } else {
-            high = mid;
+    if j == 0 {
+        while stops.get(j + 1).is_some_and(|(_, r)| r.is_zero()) {
+            j += 1;
         }
+        return stops[j].0;
     }
-
-    if low == 0 {
-        low = 1;
+    if j == stops.len() {
+        while j > 0 && stops[j].1.is_one() {
+            j -= 1;
+        }
+        return stops[j].0;
     }
```

Comment:
Since the partition checks for `< t` and the last point is required to be `1` this if statement seems to be redundant, because `1 < 1` will always return `false`.
The loop is also redundant, because we will always land at the lower end of a chunk with equal ratios. And it would panic because it would index the array at `stops.len()` when the condition is initially checked.